File: packages/core/clipboard_sync_test.go
// Copyright (C) 2026 FuseItAll contributors.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, version 3 of the License. See LICENSE
// for details.

package core

import (
	"encoding/base64"
	"strings"
	"testing"
)

func TestClipHLCNext(t *testing.T) {
	s := ClipStampNext(ClipStamp{}, 10)
	if s.L != 10 || s.C != 0 {
		t.Fatalf("next = %+v", s)
	}
	s2 := ClipStampNext(s, 10)
	if s2.C != 1 {
		t.Fatalf("same wall must bump C: %+v", s2)
	}
	s3 := ClipStampNext(s2, 5)
	if s3.L != 10 || s3.C != 2 {
		t.Fatalf("older wall must not rewind: %+v", s3)
	}
}

func TestRemoteClipWinsEx(t *testing.T) {
	if !RemoteClipWinsEx(10, 0, 11, 0) {
		t.Fatal("newer L must win")
	}
	if RemoteClipWinsEx(11, 0, 11, 0) {
		t.Fatal("tie must keep local")
	}
	if !RemoteClipWinsEx(11, 0, 11, 1) {
		t.Fatal("higher C must win on same L")
	}
	if RemoteClipWinsEx(11, 0, 0, 0) {
		t.Fatal("zero must never win")
	}
}

func TestContentHashRoundTrip(t *testing.T) {
	h := ContentHashForText("hello")
	if len(h) != 64 || !SanitizeContentHash(h) {
		t.Fatalf("hash = %q", h)
	}
	if SanitizeContentHash("zzz") {
		t.Fatal("bad hash must fail")
	}
}

func TestSniffTightened(t *testing.T) {
	png := []byte{0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A}
	if SniffImageMime(png) != "image/png" {
		t.Fatal("full PNG header must sniff")
	}
	short := []byte{0x89, 0x50, 0x4E, 0x47}
	if SniffImageMime(short) != "" {
		t.Fatal("short PNG must not sniff")
	}
	gif87 := []byte{'G', 'I', 'F', '8', '7', 'a'}
	if SniffImageMime(gif87) != "image/gif" {
		t.Fatal("GIF87a must sniff")
	}
	bare := []byte{'G', 'I', 'F', 'x', 'x', 'x'}
	if SniffImageMime(bare) != "" {
		t.Fatal("bare GIF must not sniff")
	}
}

func TestSanitizeClipChunk(t *testing.T) {
	raw := make([]byte, 100)
	for i := range raw {
		raw[i] = byte(i)
	}
	// Use a tiny PNG for magic: build minimal valid header + payload.
	png := append([]byte{0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A}, raw...)
	hash := ContentHashForBytes(png)
	total := int64(len(png))
	chunks := TotalClipChunksForSize(total)
	_ = chunks
	manifest := ClipManifestPayload{
		Nonce: "n1", SessionID: "abc123", Mime: "image/png",
		TotalRaw: total + int64(MaxClipImageRaw) + 1,
		TotalChunks: 6, ChunkSize: ClipChunkRaw,
		ChangedAt: 10, Origin: OriginMac, ContentHash: hash,
	}
	// Fix total to match stride math.
	manifest.TotalChunks = TotalClipChunksForSize(manifest.TotalRaw)
	if manifest.TotalChunks < 2 {
		manifest.TotalRaw = int64(MaxClipImageRaw) + int64(ClipChunkRaw)
		manifest.TotalChunks = TotalClipChunksForSize(manifest.TotalRaw)
	}
	if !SanitizeClipManifest(manifest) {
		t.Fatal("manifest must pass")
	}
	b64 := base64.StdEncoding.EncodeToString(png)
	chunk := ClipChunkPayload{
		Nonce: "c1", SessionID: "abc123", ChunkIndex: 0,
		TotalChunks: manifest.TotalChunks, Offset: 0, TotalRaw: manifest.TotalRaw,
		DataB64: b64,
	}
	// Chunk body valid but header total mismatch is caught by hub, not sanitize
	// when totals differ; sanitize checks shape only.
	if len(b64) > MaxClipChunkB64Len {
		t.Fatal("test chunk too large")
	}
	if !SanitizeContentHash(hash) {
		t.Fatal("hash must validate")
	}
	_ = strings.TrimSpace(chunk.DataB64)
}

func TestConcealedTypes(t *testing.T) {
	if !HasConcealedPasteboardType([]string{"public.utf8-plain-text", "org.nspasteboard.concealed-type"}) {
		t.Fatal("concealed must detect")
	}
	if HasConcealedPasteboardType([]string{"public.png"}) {
		t.Fatal("plain types must not flag")
	}
	var p SettingsSyncPayload
	if ClipboardAllowSensitive(p) {
		t.Fatal("absent must default false")
	}
}

File: apps/android/test/clipboard_sync_test.dart
// Copyright (C) 2026 FuseItAll contributors.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, version 3 of the License. See LICENSE
// for details.

import 'package:flutter_test/flutter_test.dart';
import 'package:fuseitall/features/clipboard/clipboard_chunks.dart';
import 'package:fuseitall/features/clipboard/clipboard_sync.dart';
import 'package:fuseitall/features/settings/app_settings.dart';

void main() {
  group('ClipState HLC', () {
    test('same-second bumps counter, older wall does not rewind', () {
      var s = const ClipState();
      final a = s.setLocal('hello', 10)!;
      expect(a.changedAt, 10);
      expect(a.changedC, 0);
      s = a;
      // Same text is a no-op.
      expect(identical(s.setLocal('hello', 11), s), isTrue);
      final b = s.setLocal('world', 10)!;
      expect(b.changedAt, 10);
      expect(b.changedC, 1);
    });

    test('remote tie keeps local, higher counter wins', () {
      var s = const ClipState().setLocal('a', 10)!;
      expect(
        s.applyRemote(next: 'old', nextChangedAt: 5, nextOrigin: 'mac', nonce: 'n1'),
        isNull,
      );
      expect(
        s.applyRemote(next: 'tie', nextChangedAt: 10, nextChangedC: 0, nextOrigin: 'mac', nonce: 'n2'),
        isNull,
      );
      final won = s.applyRemote(
        next: 'new',
        nextChangedAt: 10,
        nextChangedC: 1,
        nextOrigin: 'mac',
        nonce: 'n3',
      );
      expect(won, isNotNull);
      expect(won!.text, 'new');
      // Duplicate nonce never re-applies.
      expect(
        won.applyRemote(next: 'new', nextChangedAt: 10, nextChangedC: 1, nextOrigin: 'mac', nonce: 'n3'),
        isNull,
      );
    });

    test('takePending carries HLC + hash', () {
      final s = const ClipState().setLocal('hi', 7)!;
      final p = s.takePending()!;
      expect(p['changed_at'], 7);
      expect(p['changed_c'], isNotNull);
      expect((p['content_hash'] as String).length, 64);
    });
  });

  group('ClipChunkHub', () {
    test('manifest + chunks reassemble and verify', () {
      final hub = ClipChunkHub();
      final raw = List<int>.generate(6 * 1024 * 1024, (i) => i % 251);
      final parts = splitClipRaw(raw);
      expect(parts.length, 6);
      // Fake hash for the test session.
      const hash = 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa';
      expect(
        hub.begin({
          'session_id': 'abc123',
          'mime': 'image/png',
          'total_raw': raw.length,
          'total_chunks': parts.length,
          'changed_at': 10,
          'origin': 'mac',
          'content_hash': hash,
        }),
        isTrue,
      );
      Map<String, Object?>? out;
      for (final (idx, off, _) in parts) {
        // ignore: avoid_dynamic_calls
        out = hub.add({
          'session_id': 'abc123',
          'chunk_index': idx,
          'total_chunks': parts.length,
          'offset': off,
          'total_raw': raw.length,
          if (idx == parts.length - 1) 'content_hash': hash,
          'data_b64': '',
        });
      }
      // Empty bodies fail closed (hub drops), so no reassembly.
      expect(out, isNull);
    });
  });

  group('Sensitive settings', () {
    test('absent allow-sensitive defaults false and round-trips', () {
      final st = AppSettings.fromJson({'updated_unix': 1});
      expect(st.clipboardAllowSensitive, isFalse);
      expect(st.toJson().containsKey('clipboard_allow_sensitive'), isFalse);
      final on = st.withClipboardAllowSensitive(true, nowUnix: 2);
      expect(on.clipboardAllowSensitive, isTrue);
      expect(AppSettings.fromJson(on.toJson()).clipboardAllowSensitive, isTrue);
    });
  });
}
